xwiki-platform-core/xwiki-platform-web/src/main/webapp/templates/register_macros.vm

```diff
@@ -0,0 +1,219 @@
+        }
+      }
+    })
+  #set ($discard = $fields.add($field))
+##
+## The confirm password field, mandatory, must match password field, and must also be 6+ characters long.
+  #set ($field =
+    {'name': $confirmPasswordFieldName,
+      'label': $services.localization.render('core.register.passwordRepeat'),
+      'params': {
+        'type': 'password',
+        'autocomplete': 'off',
+        'size': '60'
+      },
+      'validate': {               
+        'mandatory': {
+          'failureMessage': $services.localization.render('core.validation.required.message')
+        },
+        'mustMatch': {                
+          'name': $passwordFieldName,
+          'failureMessage': $services.localization.render('xe.admin.registration.passwordMismatch')
+        },                
+        'regex': {
+          'pattern': '/.{6,}/',
+          'failureMessage': $services.localization.render('xe.admin.registration.passwordTooShort')
+        }
+      }
+    })
+  #set ($discard = $fields.add($field))
+#end
+##
+#*
+ * Generate HTML form.
+ *
+ * @param $fields The array of fields to use for generating HTML code.
+ * @param $request The request that is made by submitting the form.
+ *#
+#macro (generateHtml, $fields, $request)
+  ## Put the same values back into the fields (if is there any problem with a field from the request that is made).
+  #getParams($fields, $request)
+  <dl>
+  #foreach ($field in $fields)
+    #if ($field.get('name'))
+      #set ($fieldName = $field.get('name'))
+      #if ($field.get('label'))
+        #set ($label = $field.get('label'))
+        <dt><label for="$fieldName">$label
+        #if ($field.get('validate').get('mandatory'))
+          <span ##
+          #foreach ($entry in $fieldMandatoryStar.entrySet())
+            $entry.getKey()="$entry.getValue()" ##
+          #end
+          >$services.localization.render('core.validation.required')</span>
+        #end
+          </label>
+        </dt>
+      #end
+      ## If no tag then default tag is <input>
+      #if ($field.get('tag'))
+        #set ($tag = $field.get('tag'))
+      #else
+        #set ($tag = 'input')
+      #end
+      <dd><$tag id="$fieldName" ##
+      #set ($params = $field.get('params'))
+      ## If no name parameter is specified, then we use the field name.
+      #if (!$params.get('name'))
+        #set ($discard = $params.put('name', $fieldName))
+      #end
+      #foreach ($entry in $params.entrySet())
+        ## If a parameter is specified as '' then we don't include it.
+        #if ($entry.getValue() != '')
+          $entry.getKey()="$escapetool.xml($entry.getValue())" ##
+        #end
+      #end
+      ></$tag>
+      #if ($field.get('error'))
+        <span ##
+        #foreach ($entry in $failureMessageParams.entrySet())
+          $entry.getKey()="$entry.getValue()" ##
+        #end
+        >$field.get('error')</span>
+      #end
+      </dd>
+    #else
+      ERROR: Field with no name.
```

Comment:
Hmm, untranslated message. I know it's not part of this issue, but it would have been good to handle it. (optional)